### PR TITLE
RUN-743: Integrate EC2 Node Source with multiple (or all) regions

### PR DESCRIPTION
**Pull request by jsboak**
This PR enables a single node-source to integrate (pull in EC2's) from multiple regions by specifying multiple **Endpoints** using a comma-separated list of EC2 "endpoints" or by specifying `ALL_REGIONS`.

This is backwards compatible because the `Endpoint` field can still utilize a single entry.

To properly test, make sure the plugin retrieves nodes when the `Endpoint` field contains:
1. A single endpoint (e.g. `https://ec2.us-west-1.amazonaws.com`)
2. Multiple endpoints separated by a comma (with/without spaces after commas). Example: `https://ec2.us-west-1.amazonaws.com, https://ec2.us-east-1.amazonaws.com,https://ec2.us-west-2.amazonaws.com`
3. `ALL_REGIONS` -> This will search for (and pull in) instances from all regions that the Credentials (or IAM Role) have access to.
4. The Node Source fails and shows failure message when any of the entries in the `Endpoint` field are misspelled.

_Multiple Endpoints:_
Screen Shot 2022-03-10 at 5 50 03 PM

Screen Shot 2022-03-10 at 5 49 16 PM

_ALL_REGIONS_
Screen Shot 2022-03-14 at 7 45 31 PM

_Failure Message_
Screen Shot 2022-03-11 at 9 04 34 AM

### CODE REVIEW

**Comment by fdevans** [comment]
This may require #118 to complete successfully.

File: src/main/java/com/dtolabs/rundeck/plugin/resources/ec2/InstanceToNodeMapper.java [hunk 1]
```diff
@@ -91,21 +92,49 @@
...
             if (null != credentials) {
                 ec2 = new AmazonEC2Client(credentials, clientConfiguration);
             } else {
                 ec2 = new AmazonEC2Client(clientConfiguration);
             }
         }
-        if (null != getEndpoint()) {
-            ec2.setEndpoint(getEndpoint());
+
+        if (getEndpoint().equals("ALL_REGIONS")) {
```
Comment (qualman) [inline src/main/java/com/dtolabs/rundeck/plugin/resources/ec2/InstanceToNodeMapper.java:107]: This needs to be `if (getEndpoint() == "ALL_REGIONS") {`

**Review by qualman — CHANGES_REQUESTED** [review]
Just a few things to adjust! If you make the changes I laid out (or something similar), you end up with 3 tests that just need to be adjusted:

```
Failed to parse endpoint: Region cannot be empty
java.lang.IllegalArgumentException: Failed to parse endpoint: Region cannot be empty
	at com.dtolabs.rundeck.plugin.resources.ec2.InstanceToNodeMapper.performQuery(InstanceToNodeMapper.java:120)
	at com.dtolabs.rundeck.plugin.resources.ec2.InstanceToNodeMapperSpec.extra mapping image(InstanceToNodeMapperSpec.groovy:213)


Failed to parse endpoint: Region cannot be empty
java.lang.IllegalArgumentException: Failed to parse endpoint: Region cannot be empty
	at com.dtolabs.rundeck.plugin.resources.ec2.InstanceToNodeMapper.performQuery(InstanceToNodeMapper.java:120)
	at com.dtolabs.rundeck.plugin.resources.ec2.InstanceToNodeMapperSpec.extra mapping not calling image list(InstanceToNodeMapperSpec.groovy:254)


Failed to parse endpoint: Region cannot be empty
java.lang.IllegalArgumentException: Failed to parse endpoint: Region cannot be empty
	at com.dtolabs.rundeck.plugin.resources.ec2.InstanceToNodeMapper.performQuery(InstanceToNodeMapper.java:120)
	at com.dtolabs.rundeck.plugin.resources.ec2.InstanceToNodeMapperSpec.adding region to the node attributes(InstanceToNodeMapperSpec.groovy:293)


com.dtolabs.rundeck.plugin.resources.ec2.InstanceToNodeMapperSpec > extra mapping image FAILED
    java.lang.IllegalArgumentException at InstanceToNodeMapperSpec.groovy:213
    java.lang.IllegalArgumentException at InstanceToNodeMapperSpec.groovy:213
com.dtolabs.rundeck.plugin.resources.ec2.InstanceToNodeMapperSpec > extra mapping not calling image list FAILED
    java.lang.IllegalArgumentException at InstanceToNodeMapperSpec.groovy:254
com.dtolabs.rundeck.plugin.resources.ec2.InstanceToNodeMapperSpec > adding region to the node attributes FAILED
    java.lang.IllegalArgumentException at InstanceToNodeMapperSpec.groovy:293
17 tests completed, 3 failed
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':test'.
```

 After that, this should be good to go.

**Comment by jsboak** [comment]
@qualman want to take a look again and make sure all looks good?

**Comment by qualman** [comment]
Okay, mea culpa, I think I misread an error or something. I think it _does_ in fact need to be `getEndpoint().equals("ALL_REGIONS")`, and that's the error here. I am going to test that real quick

**Comment by qualman** [comment]
Okay, @jsboak, I pushed a change that fixes that, and I tested locally and all is well. I am going to wait for tests and then we're ready to go

**Pull request merged**